{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e3657fbf",
   "metadata": {},
   "source": [
    "# Text Extraction Pipeline\n",
    "\n",
    "## Introduction\n",
    "\n",
    "The first step in processing the input file is extracting its text content. This notebook walks you through the text extraction process along with some explanation for the reasoning behind each step. We finish with a Discussion of the benefits and limitations of this approach along with potential next steps.\n",
    "\n",
    "### Why PyMuPDF?\n",
    "\n",
    "There are several Python libraries that can extract text from a pdf including PyMuPDF, PyPDF2, Camelot, and more. After evaluating a few options and testing them on the provided document, I selected PyMuPDF for the following reasons:\n",
    "\n",
    "1. **Accurate Layout Preservation**:\n",
    "    PyMuPDF’s maintains paragraph structure and follows a natural reading order (top-down, left-to-right). This is especially important for multi-column documents like ours, since some of the other libraries I tested either scramble the text order or require hardcoded column settings, thus reducing flexibility and generalizability.\n",
    "\n",
    "2. **Rich Text Metadata**:\n",
    "    PyMuPDF provides detailed styling information for each line of text, including font family, font size, and formatting (bold, italic, etc.). Currently, the pipeline only uses font size to distinguish different parts of the article (e.g. headers vs. body text), but the additional formatting info opens the door for more advanced  processing in future iterations.\n",
    "\n",
    "3. **Performance**:\n",
    "    PyMuPDF is known for its fast text extraction, which, while not a critical factor at this prototyping stage, would be important in a production environment where efficiency and scalability matter. See: https://pymupdf.readthedocs.io/en/latest/about.html#performance\n",
    "\n",
    "## Import Libraries and Load Document"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee672296",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "from typing import List, Dict\n",
    "from time import time\n",
    "\n",
    "import pymupdf\n",
    "\n",
    "## Make sure you have placed the file in the sources directory\n",
    "doc = pymupdf.open(\"../sources/SlamonetalSCIENCE1987.pdf\")  # open a document\n",
    "\n",
    "## Get the second page for this example, as the first page is not part of the article\n",
    "page = doc[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "337f9a20",
   "metadata": {},
   "source": [
    "## Raw Text Extraction\n",
    "\n",
    "First, let's try to naively extract the text using the default \"text\" extraction option."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a61ac6b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Human Breast Cancer: Correlation of\n",
      "Relapse and Survival with Amplification\n",
      "of the HER-2lneu Oncogene\n",
      "DENNIS J. SLAMON,* GARY M. CLARK, STEVEN G. WONG, WENDY J. LEVIN,\n",
      "AxEL ULLRICH, WILLiAM L. McGuIRE\n",
      "The HER-2/neu oncoFene is a member of the erbB-like\n",
      "oncogene family, and aS related to, but distinct firom, the\n",
      "epidermal growth factor receptr. This gene has been\n",
      "shown to be amplified i human brt cancer cell lines.\n",
      "In the current study, alterations of the gene in 189\n",
      "primary human breast cancers were instigated HER-2/\n",
      "neu was found to be amplified frm 2- to\n",
      "eater than 20-\n",
      "fold in 30% ofthe tumors. Correlation ofgene amplifica-\n",
      "tion with several disease parameters was evaluated Am-\n",
      "plification of the HER-2/neu gene was a significant pre-\n",
      "dictor of both overall survival and time to relapse in\n",
      "patients with breast cancer. It retained its significance\n",
      "even when adjustments were made for other known\n",
      "prognostic factors. Moreover, HER-2/neu amplification\n",
      "had greater prognostic value than most currently used\n",
      "prognostic factors, incuding hormonal-receptor status,\n",
      "in lymph node-positive disease. These data indicate that\n",
      "thbis gene may\n",
      "lay a role in the biologic behavior and/or\n",
      "pathogenesis\n",
      "human breast cancer.\n",
      "HE EVIDENCE LINKING PROTO-ONCOGENES TO THE INDUC-\n",
      "tion or maintenance of human malignancies is largely cir-\n",
      "cumstantial, but has become increasingly compelling. This\n",
      "circumstantial evidence is derived from studies of animal models,\n",
      "tumor cell lines, and actual human tumors. Data from animal models\n",
      "and cell lines include: (i) sequence homology between human proto-\n",
      "oncogenes and the viral oncogenes oftransforming retroviruses that\n",
      "are known to be tumorigenic in some species (1, 2); (ii) transfction\n",
      "studies showing the transforming potential of proto-oncogenes m\n",
      "NIH 3T3 cells and primary embryo fibroblasts (3-5); and (iii) the\n",
      "central role of certain proto-oncogenes in tumorigenesis by chronic\n",
      "transforming retroviruses such as avian leukosis virus (6). Data from\n",
      "human tumors include: (i) increased expression of specific proto-\n",
      "oncogenes in some human malignancies (7, 8); (ii) localization of\n",
      "proto-oncogenes at or near the site of specific, tumor-associated\n",
      "chromosomal translocations (9); and (iii) amplification of proto-\n",
      "oncogenes in some human tumors (10, 11).\n",
      "Additional data linking proto-oncogenes to cell growth is their\n",
      "expression in response to certain proliferation signals (12, 13) and\n",
      "their expression during embryonic development (14, 15). More\n",
      "direct evidence comes from the fact that, of the 20 known proto-\n",
      "oncogenes, three are related to a growth factor or a growth factor\n",
      "receptor. These genes include c-sis, which is homologous to the\n",
      "9 jANUARY I987\n",
      "transforming gene ofthe simian sarcoma virus and is the 1B chain of\n",
      "platelet-derived growth factor (PDGF) (16, 17); c-fis, which is\n",
      "homologous to the transforming gene of the feline sarcoma virus\n",
      "and is closely related to the macrophage colony-stimulating factor\n",
      "receptor (CSF-1R) (18); and c-erbB, which encodes the EGF\n",
      "receptor (EGFR) and is highly homologous to the transforming\n",
      "gene of the avian erythroblastosis virus (19). The two receptor-\n",
      "related proto-oncogenes, c-fins and c-erbB, are members of the\n",
      "tyrosine-specific protein kinase family to which many proto-onco-\n",
      "genes belong.\n",
      "Recently, a novel transforming gene was identified as a result of\n",
      "transfction studies with DNA from chemically induced rat neu-\n",
      "roglioblastomas (20). This gene, called neu, was shown to be related\n",
      "to, but distinct from, the c-erbB proto-oncogene (21). By means of\n",
      "v-erbB and human EGFR as probes to screen human genomic and\n",
      "complementary DNA (cDNA) libraries, two other groups indepen-\n",
      "dently isolated human erbB-related genes that they called HER-2\n",
      "(22) and c-erbB-2 (23). Subsequent sequence analysis and chromo-\n",
      "somal mapping studies revealed all three genes (neu, c-erbB-2, and\n",
      "HER-2) to be the same (22, 24, 25). A fourth group, also using v-\n",
      "erbB as a probe, identified the same gene in a mammary carcinoma\n",
      "cell line, MAC 117, where it was found to be amplified five- to ten-\n",
      "fold (26).\n",
      "This gene, which we will call HER-2/neu, encodes a new member\n",
      "of the tyrosine kinase family; and is dosely related to, but distinct\n",
      "from, the EGFR gene (22). HER-2/neu differs from EGFR in that it\n",
      "is found on band q21 ofchromosome 17 (22, 24, 25), as compared\n",
      "to band pll.-p13 of chromosome 7, where the EGFR gene is\n",
      "located (27). Also, the HER-2/neu gene generates a messenger\n",
      "RNA (mRNA) of 4.8 kb (22), which differs from the 5.8- and 10-\n",
      "kb transcipts for the EGFR gene (28). Finally, the protein encoded\n",
      "by the HER-2/neu gene is 185,000 daltons (21), as compared to the\n",
      "170,000-dalton protein encoded by the EGFR gene. Conversely, on\n",
      "the basis ofsequence data, HER-2/neu is more closely related to the\n",
      "EGFR gene than to other members of the tyrosine kinase family\n",
      "(22). Like the EGFR protem, HER-2/neu has an extracellular\n",
      "domain, a transmembrane domain that includes two cysteine-rich\n",
      "repeat dusters, and an intracellular kinase domain (21), indicating\n",
      "D. J. Slamon, S. G. Wong, and W. J. Levin are in the Division of Hematology-\n",
      "Oncokogy, Department of Medicine and Jonsson Comprehensive Cancer Center,\n",
      "UCLA School ofMedicine, Los Angcels, CA 90024. G. AM. Clark and W. L. McGuire\n",
      "are in the Division ofOncology, Department ofMedicine, University ofTexas Health\n",
      "Science Ccnter at San Antonio, San Antonio, TX 78284. A. Utlrich\n",
      "is in the\n",
      "Deparmenet ofMolecular Biology, Gcnentech, Inc., South San Francisco, CA 94080.\n",
      "*To whom correspondence should be addressed.\n",
      "ARTICLES\n",
      "177\n",
      " on January 15, 2007 \n",
      "www.sciencemag.org\n",
      "Downloaded from \n",
      "BIOM 255 (Leffert) – Discussion Feb. 1, 2007\n",
      "\n"
     ]
    }
   ],
   "source": [
    "txt = page.get_text(option=\"text\")\n",
    "print(txt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c31bf7d",
   "metadata": {},
   "source": [
    "As you can see, the text is extracted rather cleanly and in the correct order (no issues with the columns). However, the extractor reads in some extra line breaks due to the page formatting, which sometimes splits sentences in odd places. Additionally, the extractor picks up a lot of text that is not part of the actual article, such as the journal's volume and year in the footer or the \"*To whom correspondence should be addressed\" note."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b33fa5a",
   "metadata": {},
   "source": [
    "## Rich Text Extraction\n",
    "\n",
    "This text extraction above is good enough for our retrieval and text generation pipelines, however we can fix some of the issues above by extracting some styling information about the text by using PyMuPDF's dictionary extract option and inspecting individual text blocks.\n",
    "\n",
    "Below, we compute the average font size for each text block (roughly equivalent to a paragraph) rounded to the nearest 2.5. We perform this rounding because the OCR process that produced the document led to some small variation in font size (e.g., text in the same sentence with sizes 10.1, 10.05. 10.25, etc.)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7e9eb79d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Block text: Human Breast Cancer: Correlation of Relapse and Survival with Amplification of the HER-2lneu Oncogene\n",
      "Average font size: 20.0\n",
      "\n",
      "Block text: DENNIS J. SLAMON,* GARY M. CLARK, STEVEN G. WONG, WENDY J. LEVIN, AxEL ULLRICH, WILLiAM L. McGuIRE\n",
      "Average font size: 15.0\n",
      "\n",
      "Block text: The HER-2/neu oncoFene is a member of the erbB-like oncogene family, and aS related to, but distinct firom, the epidermal growth factor receptr. This gene has been shown to be amplified i human brt cancer cell lines. In the current study, alterations of the gene in 189 primary human breast cancers were instigated HER-2/ neu was found to be amplified frm 2- to eater than 20- fold in 30% of the tumors. Correlation ofgene amplifica- tion with several disease parameters was evaluated Am- plification of the HER-2/neu gene was a significant pre- dictor of both overall survival and time to relapse in patients with breast cancer. It retained its significance even when adjustments were made for other known prognostic factors. Moreover, HER-2/neu amplification had greater prognostic value than most currently used prognostic factors, incuding hormonal-receptor status, in lymph node-positive disease. These data indicate that thbis gene may lay a role in the biologic behavior and/or pathogenesis human breast cancer.\n",
      "Average font size: 10.0\n",
      "\n",
      "Block text: HE EVIDENCE LINKING PROTO-ONCOGENES TO THE INDUC- tion or maintenance of human malignancies is largely cir- cumstantial, but has become increasingly compelling. This circumstantial evidence is derived from studies of animal models, tumor cell lines, and actual human tumors. Data from animal models and cell lines include: (i) sequence homology between human proto- oncogenes and the viral oncogenes of transforming retroviruses that are known to be tumorigenic in some species (1, 2); (ii) transfction studies showing the transforming potential of proto-oncogenes m NIH 3T3 cells and primary embryo fibroblasts (3-5); and (iii) the central role of certain proto-oncogenes in tumorigenesis by chronic transforming retroviruses such as avian leukosis virus (6). Data from human tumors include: (i) increased expression of specific proto- oncogenes in some human malignancies (7, 8); (ii) localization of proto-oncogenes at or near the site of specific, tumor-associated chromosomal translocations (9); and (iii) amplification of proto- oncogenes in some human tumors (10, 11). Additional data linking proto-oncogenes to cell growth is their expression in response to certain proliferation signals (12, 13) and their expression during embryonic development (14, 15). More direct evidence comes from the fact that, of the 20 known proto- oncogenes, three are related to a growth factor or a growth factor receptor. These genes include c-sis, which is homologous to the\n",
      "Average font size: 10.0\n",
      "\n",
      "Block text: 9 jANUARY I987\n",
      "Average font size: 7.5\n",
      "\n",
      "Block text: transforming gene of the simian sarcoma virus and is the 1B chain of platelet-derived growth factor (PDGF) (16, 17); c-fis, which is homologous to the transforming gene of the feline sarcoma virus and is closely related to the macrophage colony-stimulating factor receptor (CSF-1R) (18); and c-erbB, which encodes the EGF receptor (EGFR) and is highly homologous to the transforming gene of the avian erythroblastosis virus (19). The two receptor- related proto-oncogenes, c-fins and c-erbB, are members of the tyrosine-specific protein kinase family to which many proto-onco- genes belong. Recently, a novel transforming gene was identified as a result of transfction studies with DNA from chemically induced rat neu- roglioblastomas (20). This gene, called neu, was shown to be related to, but distinct from, the c-erbB proto-oncogene (21). By means of v-erbB and human EGFR as probes to screen human genomic and complementary DNA (cDNA) libraries, two other groups indepen- dently isolated human erbB-related genes that they called HER-2 (22) and c-erbB-2 (23). Subsequent sequence analysis and chromo- somal mapping studies revealed all three genes (neu, c-erbB-2, and HER-2) to be the same (22, 24, 25). A fourth group, also using v- erbB as a probe, identified the same gene in a mammary carcinoma cell line, MAC 117, where it was found to be amplified five- to ten- fold (26). This gene, which we will call HER-2/neu, encodes a new member of the tyrosine kinase family; and is dosely related to, but distinct from, the EGFR gene (22). HER-2/neu differs from EGFR in that it\n",
      "Average font size: 10.0\n",
      "\n",
      "Block text: is found on band q21 of chromosome 17 (22, 24, 25), as compared to band pll.-p13 of chromosome 7, where the EGFR gene is located (27). Also, the HER-2/neu gene generates a messenger RNA (mRNA) of 4.8 kb (22), which differs from the 5.8- and 10- kb transcipts for the EGFR gene (28). Finally, the protein encoded by the HER-2/neu gene is 185,000 daltons (21), as compared to the 170,000-dalton protein encoded by the EGFR gene. Conversely, on the basis of sequence data, HER-2/neu is more closely related to the EGFR gene than to other members of the tyrosine kinase family (22). Like the EGFR protem, HER-2/neu has an extracellular domain, a transmembrane domain that includes two cysteine-rich repeat dusters, and an intracellular kinase domain (21), indicating\n",
      "Average font size: 10.0\n",
      "\n",
      "Block text: D. J. Slamon, S. G. Wong, and W. J. Levin are in the Division of Hematology- Oncokogy, Department of Medicine and Jonsson Comprehensive Cancer Center, UCLA School of Medicine, Los Angcels, CA 90024. G. AM. Clark and W. L. McGuire are in the Division of Oncology, Department of Medicine, University of Texas Health Science Ccnter at San Antonio, San Antonio, TX 78284. A. Utlrich is in the Deparmenet of Molecular Biology, Gcnentech, Inc., South San Francisco, CA 94080.\n",
      "Average font size: 7.5\n",
      "\n",
      "Block text: *To whom correspondence should be addressed.\n",
      "Average font size: 7.5\n",
      "\n",
      "Block text: ARTICLES 177\n",
      "Average font size: 7.5\n",
      "\n",
      "Block text:  on January 15, 2007  www.sciencemag.org Downloaded from \n",
      "Average font size: 10.0\n",
      "\n",
      "Block text: BIOM 255 (Leffert) – Discussion Feb. 1, 2007\n",
      "Average font size: 7.5\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Read page text as a dictionary\n",
    "blocks = page.get_text(option=\"dict\", flags=11)[\"blocks\"]\n",
    "\n",
    "for b in blocks:  # iterate through the text blocks\n",
    "    font_sizes = []\n",
    "    for l in b[\"lines\"]:  # iterate through the text lines\n",
    "        for s in l[\"spans\"]:  # iterate through the text spans\n",
    "            font_sizes.append(s[\"size\"])\n",
    "    if font_sizes:\n",
    "        block_text = \" \".join(s[\"text\"] for l in b[\"lines\"] for s in l[\"spans\"])\n",
    "        print(f\"Block text: {block_text}\")\n",
    "\n",
    "        # Compute average font size of the block rounded to the nearest 2.5\n",
    "        avg_font_size = round(sum(font_sizes) / len(font_sizes) / 2.5) * 2.5\n",
    "        print(f\"Average font size: {avg_font_size}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63119ab3",
   "metadata": {},
   "source": [
    "Now we are really getting somewhere! Note that:\n",
    "- The article's title is font 20\n",
    "- The authors are in font 15\n",
    "- The body text is font 10\n",
    "- The notes, footers, and other extraneous text is font 7.5 or smaller\n",
    "\n",
    "## Complete Pipeline\n",
    "\n",
    "With that in mind, we can now make a class to process the entire document. Here is a summary of the entire process:\n",
    "1. Open the file\n",
    "2. For each page, extract the blocks along with their average font size. During this step drop any vertical text to remove the \"Downloaded from ...\" note in the righthand margin.\n",
    "3. Iterate through the blocks until you encounter a block of font size 20 (indicating the start of the article).\n",
    "4. After the start is found, continue iterating and:\n",
    "    - Add first block of size 15 to the authors\n",
    "    - Add any blocks of size 10 to the body\n",
    "    - Skip any blocks of font size <= 7.5\n",
    "    - Add any blocks of different size to a list of irregular blocks\n",
    "5. Stop when the end of the document is reached OR when you encounter another block of size 20 (indicating the start of a new article)\n",
    "6. Clean up the body text by removing line breaks that interrupt sentences, which are the result of a sentence stretching across columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3290c65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Article:\n",
    "    def __init__(\n",
    "        self,\n",
    "        path: os.PathLike,\n",
    "        title_size: int = 20,\n",
    "        author_size: int = 15,\n",
    "        body_size: int = 10,\n",
    "        note_size: int = 7.5,\n",
    "    ):\n",
    "        # Init attributes\n",
    "        self.path = path\n",
    "        self.title_size = title_size\n",
    "        self.author_size = author_size\n",
    "        self.body_size = body_size\n",
    "        self.note_size = note_size\n",
    "\n",
    "        # Init data attributes to None\n",
    "        self.title = None\n",
    "        self.authors = None\n",
    "        self.body = None\n",
    "        self.irregular_blocks = None\n",
    "\n",
    "        # Open the PDF document\n",
    "        self.doc = pymupdf.open(path)\n",
    "        self.num_pages = len(self.doc)\n",
    "\n",
    "        # Process the PDF file to extract title, authors, and body text\n",
    "        self._process_file()\n",
    "\n",
    "    def _read_page_blocks(self, page_number: int) -> List[Dict]:\n",
    "        \"\"\"\n",
    "        Extract the rich text from a specific page of the document.\n",
    "\n",
    "        Args:\n",
    "            page_number (int): The page number to read (0-indexed).\n",
    "\n",
    "        Returns:\n",
    "            List[Dict]: A list of dictionaries representing text blocks on the page.\n",
    "        \"\"\"\n",
    "\n",
    "        page = self.doc[page_number]\n",
    "        blocks = page.get_text(option=\"dict\", flags=11)[\"blocks\"]\n",
    "        return blocks\n",
    "\n",
    "    def _parse_blocks(self, blocks: List[Dict]) -> List[Dict]:\n",
    "        \"\"\"\n",
    "        Parse the text blocks to extract text and average font size.\n",
    "\n",
    "        Args:\n",
    "            blocks (List[Dict]): The list of text blocks, each containing lines and spans.\n",
    "\n",
    "        Returns:\n",
    "            List[Dict]: A list of dictionaries with block text and average font size.\n",
    "        \"\"\"\n",
    "        parsed_blocks = []\n",
    "        for b in blocks:\n",
    "\n",
    "            # Skip vertical blocks (height > 5x width)\n",
    "            x0, y0, x1, y1 = b[\"bbox\"]\n",
    "            width = x1 - x0\n",
    "            height = y1 - y0\n",
    "\n",
    "            if height > 5 * width:\n",
    "                continue\n",
    "\n",
    "            font_sizes = []\n",
    "            for l in b[\"lines\"]:\n",
    "                for s in l[\"spans\"]:\n",
    "                    font_sizes.append(s[\"size\"])\n",
    "            if font_sizes:\n",
    "                block_text = \" \".join(s[\"text\"] for l in b[\"lines\"] for s in l[\"spans\"])\n",
    "                avg_font_size = round(sum(font_sizes) / len(font_sizes) / 2.5) * 2.5\n",
    "                parsed_blocks.append({\"text\": block_text, \"size\": avg_font_size})\n",
    "        return parsed_blocks\n",
    "\n",
    "    def _parse_doc(self) -> List[Dict]:\n",
    "        \"\"\"\n",
    "        Parse the entire document to extract text blocks and their average font sizes.\n",
    "\n",
    "        Returns:\n",
    "            List[Dict]: A list of dictionaries with block text and average font size for each page.\n",
    "        \"\"\"\n",
    "        all_blocks = []\n",
    "        for page_number in range(self.num_pages):\n",
    "            blocks = self._read_page_blocks(page_number)\n",
    "            parsed_blocks = self._parse_blocks(blocks)\n",
    "            all_blocks.extend(parsed_blocks)\n",
    "        return all_blocks\n",
    "\n",
    "    def _process_file(self) -> None:\n",
    "        \"\"\"\n",
    "        Process the PDF file to extract title, authors, and body text. Ignore any text before the first headline\n",
    "        or after the second headline (if applicable).\n",
    "        \"\"\"\n",
    "        blocks = self._parse_doc()\n",
    "        start = False\n",
    "        self.body = \"\"\n",
    "        self.irregular_blocks = []\n",
    "        for block in blocks:\n",
    "            if not start and block[\"size\"] < self.title_size:\n",
    "                continue\n",
    "            elif not start and block[\"size\"] >= self.title_size:\n",
    "                start = True\n",
    "                self.title = block[\"text\"]\n",
    "                continue\n",
    "            else:\n",
    "                if block[\"size\"] <= self.note_size:\n",
    "                    continue\n",
    "                elif block[\"size\"] == self.title_size:\n",
    "                    break\n",
    "                elif block[\"size\"] == self.author_size and not self.authors:\n",
    "                    self.authors = block[\"text\"]\n",
    "                elif block[\"size\"] == self.body_size:\n",
    "                    self.body += \"\\n\" + block[\"text\"] + \"\\n\"\n",
    "                else:\n",
    "                    self.irregular_blocks.append(block)\n",
    "\n",
    "        # Remove line breaks unless they are followed by a capital letter, indicating a new sentence\n",
    "        self.body = re.sub(r'\\n(?=[^A-Z])', '', self.body.strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "48e08846",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing time: 0.03 seconds\n",
      "\n",
      "Title: Human Breast Cancer: Correlation of Relapse and Survival with Amplification of the HER-2lneu Oncogene\n",
      "Authors: DENNIS J. SLAMON,* GARY M. CLARK, STEVEN G. WONG, WENDY J. LEVIN, AxEL ULLRICH, WILLiAM L. McGuIRE\n",
      "Body:\n",
      "\n",
      "The HER-2/neu oncoFene is a member of the erbB-like oncogene family, and aS related to, but distinct firom, the epidermal growth factor receptr. This gene has been shown to be amplified i human brt cancer cell lines. In the current study, alterations of the gene in 189 primary human breast cancers were instigated HER-2/ neu was found to be amplified frm 2- to eater than 20- fold in 30% of the tumors. Correlation ofgene amplifica- tion with several disease parameters was evaluated Am- plification of the HER-2/neu gene was a significant pre- dictor of both overall survival and time to relapse in patients with breast cancer. It retained its significance even when adjustments were made for other known prognostic factors. Moreover, HER-2/neu amplification had greater prognostic value than most currently used prognostic factors, incuding hormonal-receptor status, in lymph node-positive disease. These data indicate that thbis gene may lay a role in the biologic behavior and/or pathogenesis human breast cancer.\n",
      "HE EVIDENCE LINKING PROTO-ONCOGENES TO THE INDUC- tion or maintenance of human malignancies is largely cir- cumstantial, but has become increasingly compelling. This circumstantial evidence is derived from studies of animal models, tumor cell lines, and actual human tumors. Data from animal models and cell lines include: (i) sequence homology between human proto- oncogenes and the viral oncogenes of transforming retroviruses that are known to be tumorigenic in some species (1, 2); (ii) transfction studies showing the transforming potential of proto-oncogenes m NIH 3T3 cells and primary embryo fibroblasts (3-5); and (iii) the central role of certain proto-oncogenes in tumorigenesis by chronic transforming retroviruses such as avian leukosis virus (6). Data from human tumors include: (i) increased expression of specific proto- oncogenes in some human malignancies (7, 8); (ii) localization of proto-oncogenes at or near the site of specific, tumor-associated chromosomal translocations (9); and (iii) amplification of proto- oncogenes in some human tumors (10, 11). Additional data linking proto-oncogenes to cell growth is their expression in response to certain proliferation signals (12, 13) and their expression during embryonic development (14, 15). More direct evidence comes from the fact that, of the 20 known proto- oncogenes, three are related to a growth factor or a growth factor receptor. These genes include c-sis, which is homologous to thetransforming gene of the simian sarcoma virus and is the 1B chain of platelet-derived growth factor (PDGF) (16, 17); c-fis, which is homologous to the transforming gene of the feline sarcoma virus and is closely related to the macrophage colony-stimulating factor receptor (CSF-1R) (18); and c-erbB, which encodes the EGF receptor (EGFR) and is highly homologous to the transforming gene of the avian erythroblastosis virus (19). The two receptor- related proto-oncogenes, c-fins and c-erbB, are members of the tyrosine-specific protein kinase family to which many proto-onco- genes belong. Recently, a novel transforming gene was identified as a result of transfction studies with DNA from chemically induced rat neu- roglioblastomas (20). This gene, called neu, was shown to be related to, but distinct from, the c-erbB proto-oncogene (21). By means of v-erbB and human EGFR as probes to screen human genomic and complementary DNA (cDNA) libraries, two other groups indepen- dently isolated human erbB-related genes that they called HER-2 (22) and c-erbB-2 (23). Subsequent sequence analysis and chromo- somal mapping studies revealed all three genes (neu, c-erbB-2, and HER-2) to be the same (22, 24, 25). A fourth group, also using v- erbB as a probe, identified the same gene in a mammary carcinoma cell line, MAC 117, where it was found to be amplified five- to ten- fold (26). This gene, which we will call HER-2/neu, encodes a new member of the tyrosine kinase family; and is dosely related to, but distinct from, the EGFR gene (22). HER-2/neu differs from EGFR in that itis found on band q21 of chromosome 17 (22, 24, 25), as compared to band pll.-p13 of chromosome 7, where the EGFR gene is located (27). Also, the HER-2/neu gene generates a messenger RNA (mRNA) of 4.8 kb (22), which differs from the 5.8- and 10- kb transcipts for the EGFR gene (28). Finally, the protein encoded by the HER-2/neu gene is 185,000 daltons (21), as compared to the 170,000-dalton protein encoded by the EGFR gene. Conversely, on the basis of sequence data, HER-2/neu is more closely related to the EGFR gene than to other members of the tyrosine kinase family (22). Like the EGFR protem, HER-2/neu has an extracellular domain, a transmembrane domain that includes two cysteine-rich repeat dusters, and an intracellular kinase domain (21), indicatingthat it too is likely to be a cellular receptor for an as yet unidentified ligand. As a result of the published data showing amplification of HER- 2/neu in a human mammary carcinoma cell line, and as part of an ongoing survey in our laboratory of proto-oncogene abnormalities in human tumors, we evaluated alterations of the HER-2/neu gene in a large series of human primary breast cancers. Our results show that amplification of this gene occurs relatively frequently in breast cancer, and that it is associated with disease relapse and overall patient survival. Factors that are known to be important in the prognosis of breast malignancies in individual patients include: size of the primary tumor, stage of disease at diagnosis, hormonal receptor status, and number of axillary lymph nodes involved with disease (positive nodes) (29). The current study, which was conducted in two parts, involved the evaluation of tissue from 189 separate breast malignan- cies that were part of a breast cancer study ongoing at the University of Texas, San Antonio. This cohort of tumors was of interest because considerable information was available on the majority of the specimens including size of the primary tumor, estrogen recep- tor status, progesterone receptor status, age of patient, disease stage, and status of the axillary lymph nodes. In the initial survey, tissue from 103 primary breast cancers was evaluated for alterations in the HER-2/neu gene. DNA from individual tumors was prepared as described (30), digested with Eco RI, and subjected to Southern blot analysis with a 32P-labeled HER-2/neu-1 probe, which is known to detect a 13-kb hybridizing band in human DNA (22). Examples of tumors from the initial survey are shown in Fig. 1. Of the 103 samples examined, 19 (18%) showed evidence of HER-2/neu gene amplification. The degree of amplification in individual cases was determined by dilution analysis (Fig. 2A), as well as soft laser densitometry scanning. To determine that the amount of DNA loaded in each lane was equivalent, allfilters were washed and rehybridized with a 32P-labeled arginase gene probe (31). This probe identifies a 15-kb hybridizing band on Eco RI-digested human DNA, and was selected as a control because it more appropriately assesses the relative amount andtransfer of high molecular weight species than a probe hybridizing with low molecular weight species, which transfer more readily on Southern blotting. All lanes were shown to contain equivalent amounts of high molecular weight DNA (Fig. 2B). Individual tumors were assigned to groups containing a single copy, 2 to 5 copies, 5 to 20 copies, and greater than 20 copies of the HER-2/neu gene (Fig. 1). Assignment of tumors to the various groups was donein a blinded fashion, in that they were made without knowledge of disease parameters. Analysis of the data for association between gene amplification and a number of disease parameters was then per- formed. Of 103 tumors evaluated in the initial survey, there was essentially no correlation between gene amplification and estrogen receptor status, progesterone receptor status, size of tumors, or age at diagnosis (Table 1). However, when analysis was performed for association between HER-2/neu amplification and number of posi- tive lymph nodes, a trend was noted. This analysis showed that 4/34 (11%) of patients with no involved nodes, 2/20 (10%) with 1 to 3 involved nodes, and 8/25 (32%) with >3 involved nodes had gene amplification (P = 0.11). If these data were examined by comparing 0 to 3 positive nodes versus >3 positive nodes, the correlation with gene amplification became more significant (P < 0.05). Thus, there was a significant increase in incidence of HER-2/neu gene amplifica- tion in patients with >3 axillary lymph nodes involved with disease. A multivariate regression analysis to correlate HER-2/neu amplifica- tion with various disease parameters identified the number of positive nodes as the only significant factor, either alone or in combination, to correlate with amplification. This initial study indicated that it might be possible to discrimi- nate among node-positive patients on the basis of HER-2/neu geneamplification. It is well known that the number of positive nodes is the best prognostic factor for disease recurrence and survival in patients with breast cancer (29). Given .the correlation between number of nodes positive and HER-2/neu amplification, one might predict that amplification of this gene might also have some prognostic value. No long-term follow-up data, however, were available on the 103 patients analyzed in the initial study. For this reason, a second study was conducted on 100 breast cancer samples from patients with positive axillary lymph nodes. All of the informa- tion available for the first group of 103 patients was available for these patients. In addition, relapse and survival information was available, since these cases had a median follow-up of 46 months (range 24 to 86 months). Of these 100 samples, 86 yielded sufficient DNA for study. Amplification of the HER-2/neu gene was mea- sured as in the initial survey, and examples of tumors from this study are shown (Fig. 1). Amplification was found in 34/86 (40%) of these patients. For this larger sample of node-positive patients, several statistically significant or nearly significant relationships were observed. In agreement with the preliminary survey, there was an association between number of involved lymph nodes and HER-2/ neu amplification (Table 2). In addition, the presence of gene amplification was correlated with estrogen receptor status and size of primary tumor (Table 2). Together, these two surveys yielded data on 189 patients and the association of HER-2/neu amplifica- tion with various disease parameters in the combined group is shown in Table 3. While these correlations were of interest, the strong relationshipbetween HER-2/neu amplification and nodal status (P = 0.002) indicated that information on amplification of this gene may correlate with disease behavior; that is, recurrences and survival. To test this, univariate survival analyses were performed in which amplification was compared to relapse and survival in this patient group. A total of 35 patients had a recurrence of the disease, and 29 had died at the time of the analyses. Median times to relapse and death were 62 months and 69 months, respectively. The median follow-up time forpatients still alive was 47 months, ranging from 24 to 86 months. A total of 71 of the 86 patients (83%) received some form of therapy after mastectomy: adjuvant systemic therapy alone, 47%; adjuvant systemic therapy plus local radiation, 19%; and local radiation alone, 17%. A strong and highly statistically significant correlation was found between the degree of gene amplification and both time to disease relapse (P = <0.0001) and survival (P = 0.0011) (Table 4). Moreover, when compared in univariate analyses to other parameters, amplification of HER-2/neu was found to be superior to all other prognostic factors, with the exception of the number of positive nodes (which it equaled) in predicting time to relapse and overall survival in human breast cancer (Table 4). The association between HER-2/neu amplification and relapse and survival can be illustrated graphically in actuarial survival curves (Fig. 3, A to D). While there was a somewhat shortened time to relapse and shorter overall survival in patients having any amplification of the HER-2/neu gene in their tumors (Fig. 3, A and B), the greatest differences were found when comparing patients with >5 copies of the gene to those without amplification (single copy) (Fig. 3, C and D). Patients with greater than five copies of HER-2/neu had even shorter disease-free survival times (P = 0.015) and overall survival times (P = 0.06) when compared to patients with no amplification. The phenomenon of greater gene copy number correlating with a worse prognosis has also been seen in evaluations of N-myc gene amplification in human neuroblastomas (32). To determine if amplification of HER-2/neu was independent of other known prognostic factors in predicting disease behavior, multivariate survival analyses were performed on the 86 node- positive cases. Amplification of the gene continued to be a strong prognostic factor, providing additional and independent predictive information on both time to relapse and overall survival in thesepatients, even when other prognostic factors were taken into account (Table 4). Rearrangement of the HER-2/neu gene was rare. Of the total 189 tumors evaluated, three showed evidence of rearrangement, and in two of the three cases, the rearrangement was identical (Fig. 1, cases 77 to 79). Also, two of the rearranged HER-2/neu loci were amplified (Fig. 1, cases 78 and 79). The incidence of HER-2/neu rearrangement as determined by Eco RI digestion was too small to attempt statistical correlations. To determine whether the phenomenon of amplification of HER- 21neu in breast cancer extended to related growth factor receptors, all filters were analyzed with the EGFR probe (Fig. 4). Amplifica- tion of the EGFR gene was found in 4/189 (2%) of the cases, and rearrangement of the EGFR gene was found in one of those four cases. The incidence of EGFR amplification and rearrangement was too small to attempt statistical correlation. Comparison of HER-2/ neu amplification (53/189 or 28%) with that of the EGFR gene reveals the incidence of the former to be 14 times greater than that of the latter, indicating that the phenomenon of gene amplification is not a general one for a related tyrosine kinase-specific receptor in human breast cancer. Moreover, studies examining alterations of two other tyrosine kinase-specific proto-oncogenes, abi andfes, in breast cancer did not show amplification of these genes (33). Alterations of non-tyrosine kinase-related proto-oncogenes in thesetumors have been examined. In a survey of 121 primary breast malignancies, amplification of the c-myc gene was found in 38 (32%) (34). Attempts to correlate c-myc gene amplification with stage of disease, hormonal receptor status, histopathologic grade, or axillary node metastases showed no association. There was a statisti- cally significant association between c-myc amplification and agc at diagnosis >50 years in a group of 95 of these patients (34). Data on relapse and survival were not presented in this study; however, there was no correlation between c-myc amplification and nodal status to indicate an association with disease behavior. The exact role of various proto-oncogenes in the pathogenesis of human malignancies remains unclear. One line of evidence implicat- ing abnormalities of these genes in human disease is association of their amplification with tumor progression in specific cancers. The N-myc gene is frequently amplified in human neuroblastomas and neuroblastoma cell lines (35, 36). Studies on the N-myc proto- oncogene were the first to show a direct association between abnormalities in a proto-oncogene and clinical behavior of a human tumor. N-myc amplification and expression correlate both with stage of disease and overall survival in patients with neuroblastoma (10, 32, 37). Moreover the greater the N-myc gene copy number, the worse the patient prognosis for all stages of the disease (32). Taken together, these data indicate a role for the N-myc gene in the pathogenesis of neuroblastoma (32). Neuroblastoma is a relatively rare disease with an incidence of one per 125,000 children. Carcinoma of the breast, however, is a common malignancy affecting one of every 13 women in the United States. There are 119,000 new cases per year, and approximately 40,000 women will die of the disease in 1986 (38). Current treatment decisions for individual patients are frequently based on specific prognostic parameters. The major prognostic factors for breast cancer include presence or absence of tumor in the axillary nodes, size of the primary tumor, and presence or absence of hormonal receptors (29). The current study indicates that amplifica- tion of the HER-2/neu gene is a significant predictor of both overall survival and time to relapse in node-positive patients with breast cancer. Amplification of the gene retains its prognostic significance in multivariate analysis, even when adjustments are made for other known prognostic factors. Moreover, amplification of HER-2/neu has greater prognostic value than most currently used prognostic fictors, including progesterone and estrogen receptors, and is equivalent to and independent of the best known prognosticator- number of positive lymph nodes. Finally, the degree of HER-2/neu amplification appears to have an effect on survival, with greater copy number being associated with a worse prognosis (Fig. 3, C and D). A similar phenomenon has been observed for N-myc gene amplifica- tion in human neuroblastoma (32). The potential role of HER-2/neu in the pathogenesis of breast cancer is unknown. Like N-myc, the correlation of HER-2/neu amplfication with disease progression indicates it may be an important gene in the disease process. The role of other cellreceptors in the biology of breast cancer is well established (29, 39, 40). It is easy to speculate that a gene encoding a putative growth factor receptor, when expressed in inappropriate amounts, may give a growth advantage to the cells expressing it. Alternatively, alter- ation in the gene product itself may lead to a critical change in the receptor protein. A single point mutation in the transmembrane domain of the protein encoded by the rat neu oncogene appears to be all that is necessary for the gene to gain transforming ability (41). Whether this or a similar alteration is found in the amplified HER- 21neu gene in human breast cancer will require sequence analysis of the homologous region in the amplified human gene. In addition, studies evaluating the expression of this gene at the RNA and/or protein level will prove important in determining if HER-2/neu amplification results in an expected increased gene expression. The question of amplification of HER-2/neu in metastatic as compared to primary lesions in a given patient is important. The current study utilized only primary breast tumors for analyses. It would be of interest to determine if HER-2/neu copy number is altered as the tumor metastasizes. A recent study evaluating N-myc copy number in human small cell carcinoma of the lung showed no difference between primary and metastatic lesions (11). The initial survey from the current study showed that 15% of breast cancer patients with stage I disease (node-negative) have HER-2/ncu amplification. Unfortunately, no long-term follow-up data were available for these patients. This stage I setting may be an additional group in which HER-2/neu measurements will have an impact in predicting biologic behavior of the tumor, and as a result, in design of treatment strategy. Finally, if the HER-2/neu gene product functions as a growth factor receptor that plays a role in the pathogenesis of breast cancer, identification of its ligand and development of specific antagonists could have important therapeu- tic implications.\n",
      "REFERENCES AND NOTES\n"
     ]
    }
   ],
   "source": [
    "\n",
    "start = time()\n",
    "article = Article(\"../sources/SlamonetalSCIENCE1987.pdf\")\n",
    "end = time()\n",
    "print(f\"Processing time: {end - start:.2f} seconds\\n\")\n",
    "print(f\"Title: {article.title}\")\n",
    "print(f\"Authors: {article.authors}\")\n",
    "\n",
    "print(f\"Body:\\n\\n{article.body}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e37cd52d",
   "metadata": {},
   "source": [
    "Success! \n",
    "\n",
    "The Title and Authors were extracted correctly and the body begins after the Title and ends at the REFERENCES AND NOTES (we drop the references because they are caught by our font size <= 7.5 filter and the process breaks when it hits the start of the next article on the last page).\n",
    "\n",
    "Let's see if we encountered any irregular blocks:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b416dffc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Irregular block (size 117.5): u_0 -S;0 ,-M; 60S_tJ ~~~~~~~~~~kb tN ~~~~~~~~~~~~~~~~-12\n",
      "Irregular block (size 12.5): B\n",
      "Irregular block (size 12.5): A\n",
      "Irregular block (size 12.5): B\n"
     ]
    }
   ],
   "source": [
    "for block in article.irregular_blocks:\n",
    "    print(f\"Irregular block (size {block['size']}): {block['text']}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5232e193",
   "metadata": {},
   "source": [
    "Nothing to be concerened with, looks like some figure labels.\n",
    "\n",
    "## Export (optional)\n",
    "\n",
    "Finally, let's save the article to a markdown file for human readability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b6d8e4f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "md_text = f\"# {article.title}\\n\\n\"\n",
    "md_text += f\"**Authors:** {article.authors}\\n\\n\"\n",
    "md_text += f\"## Body\\n\\n{article.body}\\n\"\n",
    "\n",
    "with open(\"article.md\", \"w\") as f:\n",
    "    f.write(md_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e7cd71c",
   "metadata": {},
   "source": [
    "## Discussion\n",
    "\n",
    "Overall, I am satisfied with the text extraction pipeline given the limitations in time and scope. The current process performs very well on the given file and should generalize well to other files with consistent styling. A summary of the benefits, limitations, and potential improvements is given below.\n",
    "\n",
    "### Benefits\n",
    "\n",
    "- **Clean Text**: PyMuPDF extract text cleanly, with minimal missing/misidentified characters.\n",
    "- **Accurrate Layout**: PyMuPDF extracts text in the correct reading order (top-down, left-right) without the user hardcoding the documents format (e.g., number of columns or rows).\n",
    "- **Classify Text from Styling**: use the text styling (font size and orientation) to classify data as article, author, body, or extraneous. \n",
    "- **Performant**: article processing time takes approximately 0.03 seconds which is about 0.005 seconds per page.\n",
    "- **Interpretable & Customizable**: unlike black-box 3rd party services, we can control each step of the process\n",
    "\n",
    "### Limitations\n",
    "\n",
    "- **Format Fragility**: This approach assumes consistent formatting across articles. While articles in the same issue of a journal likely have the same styling, different journals may use different styles or journals may change their styling over time.\n",
    "- **Brittle Text Classification**: we only use font size to classify each block of text, ignoring any semantic understanding. This may lead to issues if the formatting is not consistent within the article, for instance if a Figure label is also Size 10.\n",
    "\n",
    "### Potential Future Improvements \n",
    "- **Autodetect Font-Sizes to Adapt to Different Styles**: dynamically determine the different font size cutoffs for body, authors, or titles\n",
    "- **Incorporate Semantic Understanding to Text Classification**: for example, author block should not only be of the right size but must contain one or more person names, validated using NER techniques.\n",
    "- **Track Section Headers or other Structure**: for long articles with mutliple sections (Background, Methods, etc.) capture the sections so we can better contextualize chunks in down stream processes.\n",
    "- **Parellelization**: the intial processing of each page individually could be parallelized once speed/scalability becomes a concern."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c84185c",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
